proposals/massive_context_refactor.cpp

```diff
@@ -0,0 +1,381 @@
+            u32,
+            float,
+            undefined
+};
+// Need to figure out way of defering reference until after lefthand assignment
+class IntegralValue {
+    void* p;
+    public:
+
+        // Explicit
+        IntegralValue(void* p) : p(p) {}
+        IntegralValue(const uint8_t& u) { *reinterpret_cast<uint8_t*>(p) = u; }
+        IntegralValue(const int8_t& u) { *reinterpret_cast<int8_t*>(p) = u; }
+        IntegralValue(const uint16_t& u) { *reinterpret_cast<uint16_t*>(p) = u; }
+        IntegralValue(const int16_t& u) { *reinterpret_cast<int16_t*>(p) = u; }
+        IntegralValue(const uint32_t& u) { *reinterpret_cast<uint32_t*>(p) = u; }
+        IntegralValue(const int32_t& u) { *reinterpret_cast<int32_t*>(p) = u; }
+
+        //IntegralValue& operator=(void* _p) { p = _p; }
+
+        operator uint8_t   ( ) const { return static_cast<uint8_t> ( *reinterpret_cast<uint8_t*>  ( p)); }
+        operator uint8_t&  ( )       { return static_cast<uint8_t&> ( *reinterpret_cast<uint8_t*>  ( p)); }
+        operator int8_t    ( ) const { return static_cast<int8_t> ( *reinterpret_cast<int8_t*>   ( p)); }
+        operator int8_t&   ( )       { return static_cast<int8_t&> ( *reinterpret_cast<int8_t*>   ( p)); }
+        
+        operator uint16_t  ( ) const { return static_cast<uint16_t> ( *reinterpret_cast<uint16_t*> ( p)); }
+        operator uint16_t& ( )       { return static_cast<uint16_t&> ( *reinterpret_cast<uint16_t*> ( p)); }
+        operator int16_t   ( ) const { return static_cast<int16_t> ( *reinterpret_cast<int16_t*>  ( p)); }
+        operator int16_t&  ( )       { return static_cast<int16_t&> ( *reinterpret_cast<int16_t*>  ( p)); }
+
+        operator uint32_t  ( ) const { return static_cast<uint32_t> ( *reinterpret_cast<uint32_t*> ( p)); }
+        operator uint32_t& ( )       { return static_cast<uint32_t&> ( *reinterpret_cast<uint32_t*> ( p)); }
+        operator int32_t   ( ) const { return static_cast<int32_t> ( *reinterpret_cast<int32_t*>  ( p)); }
+        operator int32_t&  ( )       { return static_cast<int32_t&> ( *reinterpret_cast<int32_t*>  ( p)); }
+
+        operator float   ( ) const { return static_cast<float> ( *reinterpret_cast<float*>  ( p)); }
+        operator float&  ( )       { return static_cast<float&> ( *reinterpret_cast<float*>  ( p)); }
+
+};
+
+// template<typename Allocator=utensor::DefaultTensorMetaDataAllocator>
+class TensorBase {
+public:
+    TensorBase(){
+        utensor::Context::get_default_context().register(*this);
+    }
+    //virtual ~TensorBase(){}
+
+    // Allocate the tensor metadata on a different heap from the data scratch pads
+
+    void* operator new(size_t sz) { 
+        void* p = utensor::Context::DefaultTensorMetaDataAllocator::allocate(sz); 
+        return p;
+    }
+    void operator delete(void* p) {
+        utensor::Context::DefaultTensorMetaDataAllocator::deallocate(p);
+    }
+};
+
+class TensorInterface : public TensorBase {
+    protected:
```

Comment:
Does this have to be `protected`? Performance critical code requires direct-linear-index access, ie, raw pointers.

Reply:
Yes, don't think of the TensorInterface as a public interface (Java style) think of it as the contract between the uTensor core and the Tensors. For clarity, the optimizable operators subclass the FastOperator empty type to gain access to direct memory blocks (which may need to be paged in)